File: mwflow/web/js/nucleo/litevna.js
/* Driver do LiteVNA64 — gêmeo de `mwflow/vna/litevna.py`.
 *
 * A diferença de fundo entre os dois: lá as leituras da serial são bloqueantes
 * dentro de uma thread; aqui são `await` dentro do laço de eventos do
 * navegador. A ordem das operações é a mesma, e as duas armadilhas do
 * aparelho continuam tratadas do mesmo jeito:
 *
 * 1. `LE_FIFO` tem contagem de UM byte. Acima de 255 valores é obrigatório ler
 *    em blocos. O driver de onde este veio não fazia isso e devolvia lixo.
 * 2. Os vetores nascem a cada varredura, e não uma vez só. Um vetor
 *    reaproveitado não redimensiona quando `pontos` muda.
 */

"use strict";

import * as p from "./protocolo.js";
import { complexo } from "./num.js";

export class ErroVNA extends Error {}

/** A aquisição foi interrompida a pedido, entre dois blocos da FIFO.
 *
 *  Não é erro. Uma varredura abortada é descartada inteira: entregar meia
 *  varredura corromperia o ajuste em silêncio. */
export class Abortado extends Error {}

const espera = (ms) => new Promise((ok) => setTimeout(ok, ms));

export class LiteVNA {
  static nome = "LiteVNA64";

  constructor(porta, bloco = p.BLOCO_PADRAO, tentativas = 3) {
    this.porta = porta;
    this.bloco = Math.min(bloco, p.MAX_POR_LEITURA);
    this.tentativas = tentativas;
    this.simulado = !!(porta && porta.simulado);
    // Bandeira opcional. Quando o motor a levanta, a leitura em curso para no
    // fim do bloco atual, e não no fim da varredura. Com 1024 pontos a
    // diferença é entre 0,1 s e 3,4 s de espera para trocar de banda.
    this.abortar = { ativo: false };
    this.inicioHz = 0;
    this.passoHz = 0;
    this.pontos = 0;
    this.valoresPorF = 1;
  }

  // ------------------------------------------------------- ciclo de vida
  async abre() {
    await this.porta.abre();
    await this.ressincroniza();
    return this;
  }

  async fecha() {
    if (this.porta) await this.porta.fecha();
  }

  get aberto() {
    return !!(this.porta && this.porta.aberta);
  }

  // --------------------------------------------------------- baixo nível
  /** Devolve o aparelho a um estado conhecido depois de um erro.
   *
   *  Oito NOPs fecham qualquer comando pela metade; o descarte da entrada joga
   *  fora os bytes órfãos que ainda estavam a caminho. No Linux o
   *  ModemManager sonda portas CDC-ACM com comandos AT logo depois do plugue,
   *  e esses bytes caem no interpretador binário do LiteVNA. */
  async ressincroniza() {
    this.porta.limpaEntrada();
    await this.porta.escreve(p.cmdSincroniza());
    await espera(50);
    this.porta.limpaEntrada();
  }

  async _leReg(addr, n) {
    await this.porta.escreve(p.cmdLe(addr, n));
    const d = await this.porta.leExato(n, 1000);
    if (d.length !== n) {
      throw new ErroVNA("registrador 0x" + addr.toString(16)
        + ": li " + d.length + " de " + n + " bytes");
    }
    let v = 0;
    for (let i = n - 1; i >= 0; i--) v = v * 256 + d[i];
    return v;
  }

  /** Confirma que do outro lado há mesmo um aparelho da família V2. */
  async confere() {
    this.porta.limpaEntrada();
    await this.porta.escreve(Uint8Array.from([p.INDICATE]));
    const r = await this.porta.leExato(1, 1000);
    if (!r.length) {
      throw new ErroVNA("INDICATE sem resposta — o aparelho fala o protocolo "
        + "de texto? (o LiteVNA usa o binário V2)");
    }
    if (r[0] !== p.RESPOSTA_INDICATE_V2) {
      throw new ErroVNA("INDICATE devolveu 0x" + r[0].toString(16)
        + ", esperava 0x" + p.RESPOSTA_INDICATE_V2.toString(16));
    }
    return true;
  }

  /** Identificação do aparelho. */
  async info() {
    await this.abre();
    await this.confere();
    return {
      porta: this.simulado ? "simulado" : "WebSerial",
      variante: this.simulado ? "simulado" : await this._leReg(p.REG_VARIANTE, 1),
      protocolo: await this._leReg(p.REG_PROTOCOLO, 1),
      hardware: await this._leReg(p.REG_HARDWARE, 1),
      firmware: (await this._leReg(p.REG_FW_MAIOR, 1)) + "."
        + (await this._leReg(p.REG_FW_MENOR, 1)),
    };
  }

  // -------------------------------------------------------- configuração
  /** Escreve os registradores de varredura e limpa a FIFO.
   *
   *  Ler de volta 0x00/0x10/0x20 devolve zero neste firmware. Não é falha:
   *  esses registradores só valem depois que o host escreve. Por isso o estado
   *  fica guardado aqui, e não no aparelho. */
  async configura(inicioHz, passoHz, pontos, valoresPorF = 1) {
    pontos = Math.round(pontos);
    if (!(pontos >= 1 && pontos <= p.PONTOS_MAX)) {
      throw new RangeError("pontos fora de 1.." + p.PONTOS_MAX);
    }
    const fim = inicioHz + passoHz * (pontos - 1);
    for (const f of [inicioHz, fim]) {
      if (!(f >= p.F_MIN_HZ && f <= p.F_MAX_HZ)) {
        throw new RangeError(
          (f / 1e6).toFixed(3) + " MHz fora da faixa do LiteVNA64 ("
          + (p.F_MIN_HZ / 1e3) + " kHz a " + (p.F_MAX_HZ / 1e9).toFixed(3)
          + " GHz)");
      }
    }
    this.inicioHz = inicioHz;
    this.passoHz = passoHz;
    this.pontos = pontos;
    this.valoresPorF = valoresPorF;

    await this.porta.escreve(p.cmdVarredura(inicioHz, passoHz, pontos,
                                            valoresPorF));
    // o aparelho reinicia a varredura; dá um tempo antes de limpar a FIFO
    await espera(50);
    await this.limpaFifo();
    return this;
  }

  /** Configura por início e fim, em vez de início e passo. */
  async defineVarredura(inicioHz, fimHz, pontos, valoresPorF = 1) {
    return this.configura(inicioHz, p.passoDe(inicioHz, fimHz, pontos), pontos,
                          valoresPorF);
  }

  /** Modo de frequência fixa: passo nulo.
   *
   *  Com o passo em zero todos os pontos caem na mesma frequência, e cada
   *  registro da FIFO vira uma amostra no tempo. É o que alimenta a tela do
   *  osciloscópio. */
  async defineCw(fHz, bloco = null, valoresPorF = 1) {
    return this.configura(fHz, 0, bloco || this.bloco, valoresPorF);
  }

  async limpaFifo() {
    this.porta.limpaEntrada();
    await this.porta.escreve(p.cmdLimpaFifo());
  }

  frequencias() {
    return p.gradeF(this.inicioHz, this.passoHz, this.pontos);
  }

  // ------------------------------------------------------------ aquisição
  /** Folga generosa: 2 s de partida mais 20 ms por valor. */
  _prazoDe(n) {
    return 2000 + 20 * n * Math.max(1, this.valoresPorF);
  }

  async _leBloco(n) {
    await this.porta.escreve(p.cmdLeFifo(n));
    const esperado = n * p.BYTES_POR_VALOR;
    const buf = await this.porta.leExato(esperado, this._prazoDe(n));
    if (buf.length !== esperado) {
      throw new ErroVNA("FIFO curta: " + buf.length + " de " + esperado
        + " bytes (" + n + " valores)");
    }
    return p.analisaFifo(buf);
  }

  /** Lê `total` valores em blocos, respeitando o limite de 1 byte. */
  async _leN(total) {
    const idx = new Int32Array(total);
    const s11 = complexo(total);
    const s21 = complexo(total);
    let posto = 0;
    while (posto < total) {
      if (this.abortar && this.abortar.ativo) {
        throw new Abortado("leitura interrompida com " + posto + " de " + total);
      }
      const n = Math.min(this.bloco, total - posto);
      const r = await this._leBloco(n);
      idx.set(r.idx, posto);
      s11.re.set(r.s11.re, posto); s11.im.set(r.s11.im, posto);
      s21.re.set(r.s21.re, posto); s21.im.set(r.s21.im, posto);
      posto += n;
    }
    return { idx: idx, s11: s11, s21: s21 };
  }

  /** Executa uma leitura; em falha, ressincroniza e tenta de novo. */
  async _comRetentativa(funcao) {
    let ultimo = null;
    for (let t = 0; t < this.tentativas; t++) {
      try {
        return await funcao();
      } catch (e) {
        if (e instanceof Abortado) throw e;
        if (!(e instanceof ErroVNA)) throw e;
        ultimo = e;
        if (t + 1 < this.tentativas) {
          await this.ressincroniza();
          await this.limpaFifo();
          await espera(100);
        }
      }
    }
    throw new ErroVNA("falhou após " + this.tentativas + " tentativas: "
      + (ultimo && ultimo.message));
  }

  /**
   * Uma varredura completa. Devolve `{f, s11, s21}`.
   *
   * A varredura do aparelho é cíclica. Ao ler `pontos` valores a partir de um
   * lugar qualquer do ciclo, cada índice de frequência aparece uma vez — por
   * isso o resultado é espalhado pelo `freqIndex`, e não pela ordem de
   * chegada. Se algum índice faltar, o ponto sai NaN em vez de vir de outra
   * varredura.
   *
   * Com `valoresPorF` acima de 1 é preciso ler `pontos · valoresPorF`
   * registros e promediar por índice: a sonda mostrou que o aparelho emite um
   * registro por valor e NÃO promedia por dentro.
   */
  async varre() {
    const n = this.pontos;
    const v = Math.max(1, this.valoresPorF);
    const r = await this._comRetentativa(async () => {
      await this.limpaFifo();
      return this._leN(n * v);
    });

    const conta = new Float64Array(n);
    const s11 = complexo(n);
    const s21 = complexo(n);
    for (let k = 0; k < r.idx.length; k++) {
      const i = r.idx[k];
      if (i < 0 || i >= n) continue;
      conta[i] += 1;
      s11.re[i] += r.s11.re[k]; s11.im[i] += r.s11.im[k];
      s21.re[i] += r.s21.re[k]; s21.im[i] += r.s21.im[k];
    }
    for (let i = 0; i < n; i++) {
      const c = conta[i] || NaN;      // 0/0 vira NaN, e não divisão por zero
      s11.re[i] /= c; s11.im[i] /= c;
      s21.re[i] /= c; s21.im[i] /= c;
    }
    return { f: this.frequencias(), s11: s11, s21: s21 };
  }

  /**
   * Média de `m` varreduras. Devolve `{f, s11, s21, desvio}`.
   *
   * A média é feita aqui, e não pelo registrador `valoresPorF` do aparelho:
   * assim o número de varreduras somadas é conhecido e o desvio sai junto, que
   * é o que a tela de calibração precisa.
   */
  async varreMedia(m = 1) {
    m = Math.max(1, Math.round(m));
    const primeira = await this.varre();
    if (m === 1) {
      return { f: primeira.f, s11: primeira.s11, s21: primeira.s21,
               desvio: new Float64Array(primeira.f.length) };
    }
    const n = primeira.f.length;
    const acc11 = complexo(n), acc21 = complexo(n);
    const c11 = new Float64Array(n), c21 = new Float64Array(n);
    const mods = [];
    // Média que ignora ponto sem dado, como o `np.nanmean` do lado do Python.
    // Um único NaN numa das varreduras não pode apagar o ponto na média.
    const soma = (dst, cnt, z) => {
      for (let i = 0; i < n; i++) {
        if (!Number.isFinite(z.re[i]) || !Number.isFinite(z.im[i])) continue;
        dst.re[i] += z.re[i]; dst.im[i] += z.im[i]; cnt[i] += 1;
      }
    };
    const modulo = (z) => {
      const v = new Float64Array(n);
      for (let i = 0; i < n; i++) v[i] = Math.hypot(z.re[i], z.im[i]);
      return v;
    };
    soma(acc11, c11, primeira.s11); soma(acc21, c21, primeira.s21);
    mods.push(modulo(primeira.s21));
    for (let k = 1; k < m; k++) {
      const r = await this.varre();
      soma(acc11, c11, r.s11); soma(acc21, c21, r.s21);
      mods.push(modulo(r.s21));
    }
    for (let i = 0; i < n; i++) {
      const a = c11[i] || NaN, b = c21[i] || NaN;
      acc11.re[i] /= a; acc11.im[i] /= a;
      acc21.re[i] /= b; acc21.im[i] /= b;
    }
    const desvio = new Float64Array(n);
    for (let i = 0; i < n; i++) {
      let s = 0, q = 0, c = 0;
      for (let k = 0; k < m; k++) {
        const v = mods[k][i];
        if (Number.isFinite(v)) { s += v; q += v * v; c++; }
      }
      desvio[i] = c ? Math.sqrt(Math.max(0, q / c - (s / c) * (s / c))) : NaN;
    }
    return { f: primeira.f, s11: acc11, s21: acc21, desvio: desvio };
  }

  /**
   * `n` amostras no tempo, na frequência fixa já configurada.
   *
   * Em modo CW o `freqIndex` não distingue nada — todos os pontos estão na
   * mesma frequência. Aqui vale a ordem de chegada, e não o índice.
   */
  async amostrasCw(n) {
    const largura = Math.abs(this.passoHz) * Math.max(1, this.pontos - 1);
    if (largura > 1e-6 * Math.max(this.inicioHz, 1)) {
      throw new ErroVNA("amostrasCw exige frequência fixa: a grade atual cobre "
        + largura.toFixed(3) + " Hz. Chame defineCw antes.");
    }
    const r = await this._comRetentativa(() => this._leN(n));
    return { s11: r.s11, s21: r.s21 };
  }
}

File: mwflow/vna/litevna.py
"""Driver do LiteVNA64 sobre o protocolo binário NanoVNA-V2.

Base: `repos-independentes/ndtmwscanner/src/hardware/nanovna_driver.py`, com
dois defeitos corrigidos:

1. Aquele driver escrevia `points` no byte de contagem do comando LE_FIFO. Como
   a contagem tem um byte só, qualquer varredura acima de 255 pontos devolvia
   lixo ou travava. Aqui a leitura é em blocos.
2. `sweepData` era criado uma vez e não redimensionava quando `points` mudava.
   Aqui os vetores nascem a cada varredura.

Uso pela linha de comando:

    python3 -m mwflow.vna.litevna --info
    python3 -m mwflow.vna.litevna --varre 50e6 3e9 101
    python3 -m mwflow.vna.litevna --varre 1e9 2e9 1024
    python3 -m mwflow.vna.litevna --cw 1.5e9 --n 500
"""

import sys
import time

import numpy as np
import serial

from mwflow.caminhos import PORTA_PADRAO
from mwflow.vna import protocolo as p


class ErroVNA(Exception):
    """Falha de comunicação com o aparelho."""


class Abortado(Exception):
    """A aquisição foi interrompida a pedido, entre dois blocos da FIFO.

    Não é erro. Uma varredura abortada é descartada inteira: entregar meia
    varredura corromperia o ajuste em silêncio.
    """


class LiteVNA:
    """Dono exclusivo de uma porta serial do LiteVNA.

    Um único objeto deste tipo pode existir por porta. O motor de aquisição
    mantém esse objeto dentro de uma thread só, porque o pyserial bloqueia e o
    aparelho não aceita dois interlocutores.
    """

    nome = "LiteVNA64"

    def __init__(self, porta=None, bloco=p.BLOCO_PADRAO, tentativas=3):
        self.porta = porta or PORTA_PADRAO
        self.bloco = min(int(bloco), p.MAX_POR_LEITURA)
        self.tentativas = int(tentativas)
        self.serial = None
        # Evento opcional. Quando o motor o levanta, a leitura em curso para
        # no fim do bloco atual, e não no fim da varredura. Com 1024 pontos a
        # diferença é entre 0,1 s e 3,4 s de espera para trocar de banda.
        self.abortar = None
        self.simulado = False
        # estado da varredura configurada
        self.inicio_hz = 0.0
        self.passo_hz = 0.0
        self.pontos = 0
        self.valores_por_f = 1

    # ------------------------------------------------------------ ciclo de vida
    def abre(self):
        """Abre a porta com posse exclusiva e ressincroniza.

        `exclusive=True` pede TIOCEXCL ao núcleo: um segundo processo que tente
        abrir a mesma porta falha na hora, em vez de os dois disputarem a FIFO.

        A ressincronização na abertura não é zelo excessivo. O ModemManager
        está ativo nesta máquina e marca este aparelho como candidato
        (`ID_MM_CANDIDATE=1`); ele sonda portas CDC-ACM com comandos AT logo
        depois do plugue. Esses bytes caem no interpretador binário do LiteVNA
        e podem deixá-lo no meio de um comando.
        """
        if self.serial is None:
            try:
                self.serial = serial.Serial(self.porta, 115200, timeout=2.0,
                                            exclusive=True, rtscts=False,
                                            dsrdtr=False)
            except serial.SerialException as e:
                raise ErroVNA("não abriu %s: %s" % (self.porta, e)) from e
            self.ressincroniza()
        return self

    def fecha(self):
        if self.serial is not None:
            try:
                self.serial.close()
            finally:
                self.serial = None

    def __enter__(self):
        return self.abre()

    def __exit__(self, *exc):
        self.fecha()
        return False

    @property
    def aberto(self):
        return self.serial is not None and self.serial.is_open

    # ------------------------------------------------------------- baixo nível
    def ressincroniza(self):
        """Devolve o aparelho a um estado conhecido depois de um erro.

        Oito NOPs fecham qualquer comando pela metade; o descarte da entrada
        joga fora os bytes órfãos que ainda estavam a caminho.
        """
        s = self.serial
        s.reset_input_buffer()
        s.write(p.cmd_sincroniza())
        s.flush()
        time.sleep(0.05)
        s.reset_input_buffer()

    def _le_reg(self, addr, n):
        s = self.serial
        s.timeout = 1.0
        s.write(p.cmd_le(addr, n))
        d = s.read(n)
        if len(d) != n:
            raise ErroVNA("registrador 0x%02X: li %d de %d bytes"
                          % (addr, len(d), n))
        return int.from_bytes(d, "little")

    def confere(self):
        """Confirma que do outro lado há mesmo um aparelho da família V2."""
        s = self.serial
        s.timeout = 1.0
        s.reset_input_buffer()
        s.write(bytes([p.INDICATE]))
        r = s.read(1)
        if not r:
            raise ErroVNA("INDICATE sem resposta em %s — o aparelho fala o "
                          "protocolo de texto? (LiteVNA usa o binário V2)"
                          % self.porta)
        if r[0] != p.RESPOSTA_INDICATE_V2:
            raise ErroVNA("INDICATE devolveu 0x%02X, esperava 0x%02X"
                          % (r[0], p.RESPOSTA_INDICATE_V2))
        return True

    def info(self):
        """Identificação do aparelho."""
        self.abre()
        self.confere()
        return dict(
            porta=self.porta,
            variante=self._le_reg(p.REG_VARIANTE, 1),
            protocolo=self._le_reg(p.REG_PROTOCOLO, 1),
            hardware=self._le_reg(p.REG_HARDWARE, 1),
            firmware="%d.%d" % (self._le_reg(p.REG_FW_MAIOR, 1),
                                self._le_reg(p.REG_FW_MENOR, 1)),
        )

    # ------------------------------------------------------------ configuração
    def configura(self, inicio_hz, passo_hz, pontos, valores_por_f=1):
        """Escreve os registradores de varredura e limpa a FIFO.

        Ler de volta 0x00/0x10/0x20 devolve zero neste firmware. Não é falha:
        esses registradores só valem depois que o host escreve. Por isso o
        estado fica guardado aqui, não no aparelho.
        """
        self.abre()
        pontos = int(pontos)
        if not 1 <= pontos <= p.PONTOS_MAX:
            raise ValueError("pontos fora de 1..%d" % p.PONTOS_MAX)
        fim = inicio_hz + passo_hz * (pontos - 1)
        for f in (inicio_hz, fim):
            if not (p.F_MIN_HZ <= f <= p.F_MAX_HZ):
                raise ValueError("%.6g Hz fora da faixa do LiteVNA64 "
                                 "(%.0f Hz a %.3f GHz)"
                                 % (f, p.F_MIN_HZ, p.F_MAX_HZ / 1e9))

        self.inicio_hz = float(inicio_hz)
        self.passo_hz = float(passo_hz)
        self.pontos = pontos
        self.valores_por_f = int(valores_por_f)

        self.serial.write(p.cmd_varredura(inicio_hz, passo_hz, pontos,
                                          valores_por_f))
        self.serial.flush()
        # o aparelho reinicia a varredura; dá um tempo antes de limpar a FIFO
        time.sleep(0.05)
        self.limpa_fifo()
        return self

    def define_varredura(self, inicio_hz, fim_hz, pontos, valores_por_f=1):
        """Configura por início e fim, em vez de início e passo."""
        return self.configura(inicio_hz, p.passo_de(inicio_hz, fim_hz, pontos),
                              pontos, valores_por_f)

    def define_cw(self, f_hz, bloco=None, valores_por_f=1):
        """Modo de frequência fixa: passo nulo.

        Com o passo em zero todos os pontos caem na mesma frequência, e cada
        registro da FIFO vira uma amostra no tempo. É o que alimenta a tela do
        osciloscópio.
        """
        n = int(bloco or self.bloco)
        return self.configura(f_hz, 0, n, valores_por_f)

    def limpa_fifo(self):
        self.serial.reset_input_buffer()
        self.serial.write(p.cmd_limpa_fifo())
        self.serial.flush()

    def frequencias(self):
        return p.grade(self.inicio_hz, self.passo_hz, self.pontos)

    # --------------------------------------------------------------- aquisição
    def _timeout_de(self, n):
        """Folga generosa: 2 s de partida mais 20 ms por valor.

        Vinte milissegundos por valor é cerca de cinco vezes o tempo medido
        (270 valores por segundo). A folga cobre o primeiro bloco depois de uma
        reconfiguração, quando o aparelho ainda está reiniciando a varredura.
        """
        return 2.0 + 0.02 * n * max(1, self.valores_por_f)

    def _le_bloco(self, n):
        """Lê exatamente `n` valores da FIFO. Levanta ErroVNA se vier curto."""
        s = self.serial
        s.timeout = self._timeout_de(n)
        s.write(p.cmd_le_fifo(n))
        esperado = n * p.BYTES_POR_VALOR
        buf = s.read(esperado)
        if len(buf) != esperado:
            raise ErroVNA("FIFO curta: %d de %d bytes (%d valores)"
                          % (len(buf), esperado, n))
        return p.analisa_fifo(buf)

    def _le_n(self, total):
        """Lê `total` valores em blocos, respeitando o limite de 1 byte."""
        idxs, s11s, s21s = [], [], []
        restam = total
        while restam > 0:
            if self.abortar is not None and self.abortar.is_set():
                raise Abortado("leitura interrompida com %d de %d valores"
                               % (total - restam, total))
            n = min(self.bloco, restam)
            i, a, b = self._le_bloco(n)
            idxs.append(i)
            s11s.append(a)
            s21s.append(b)
            restam -= n
        return (np.concatenate(idxs), np.concatenate(s11s),
                np.concatenate(s21s))

    def _com_retentativa(self, funcao):
        """Executa uma leitura; em falha, ressincroniza e tenta de novo."""
        ultimo = None
        for tentativa in range(self.tentativas):
            try:
                return funcao()
            except ErroVNA as e:
                ultimo = e
                if tentativa + 1 < self.tentativas:
                    self.ressincroniza()
                    self.limpa_fifo()
                    time.sleep(0.1)
        raise ErroVNA("falhou após %d tentativas: %s" % (self.tentativas, ultimo))

    def varre(self):
        """Uma varredura completa. Devolve (f, s11, s21).

        A varredura do aparelho é cíclica. Ao ler `pontos` valores a partir de
        um lugar qualquer do ciclo, cada índice de frequência aparece uma vez —
        por isso o resultado é espalhado pelo `freqIndex`, e não pela ordem de
        chegada. Se algum índice faltar, o ponto sai NaN em vez de vir de outra
        varredura.

        Com `valores_por_f` acima de 1 é preciso ler `pontos * valores_por_f`
        registros e promediar por índice: a sonda mostrou que o aparelho emite
        um registro por valor e não promedia por dentro. Ler só `pontos`
        registros devolveria uma fração da varredura.
        """
        n = self.pontos
        v = max(1, self.valores_por_f)

        def leitura():
            self.limpa_fifo()
            return self._le_n(n * v)

        idx, s11, s21 = self._com_retentativa(leitura)

        val = (idx >= 0) & (idx < n)
        idx, s11, s21 = idx[val], s11[val], s21[val]

        conta = np.bincount(idx, minlength=n).astype(float)
        soma11 = (np.bincount(idx, weights=s11.real, minlength=n)
                  + 1j * np.bincount(idx, weights=s11.imag, minlength=n))
        soma21 = (np.bincount(idx, weights=s21.real, minlength=n)
                  + 1j * np.bincount(idx, weights=s21.imag, minlength=n))

        vazio = conta == 0
        conta[vazio] = np.nan          # 0/0 -> NaN, e não divisão por zero
        return self.frequencias(), soma11 / conta, soma21 / conta

    def varre_media(self, m=1):
        """Média de `m` varreduras. Devolve (f, s11, s21, desvio_s21).

        A média é feita aqui, no software, e não pelo registrador
        `valores_por_f` do aparelho: assim o número de varreduras somadas é
        conhecido e o desvio sai junto, o que a tela de calibração precisa.
        """
        m = max(1, int(m))
        f, a, b = self.varre()
        if m == 1:
            return f, a, b, np.zeros_like(np.abs(b))
        A = [a]
        B = [b]
        for _ in range(m - 1):
            _, x, y = self.varre()
            A.append(x)
            B.append(y)
        A = np.array(A)
        B = np.array(B)
        return (f, np.nanmean(A, axis=0), np.nanmean(B, axis=0),
                np.nanstd(np.abs(B), axis=0))

    def amostras_cw(self, n):
        """`n` amostras no tempo, na frequência fixa já configurada.

        Em modo CW o `freqIndex` não distingue nada — todos os pontos estão na
        mesma frequência. Aqui vale a ordem de chegada, não o índice.
        """
        # O passo tem de ser desprezível diante da frequência. Zero é o caso
        # normal; 1 Hz em 1,5 GHz é a alternativa de micro-varredura, que
        # existe caso algum firmware recuse passo nulo.
        largura = abs(self.passo_hz) * max(1, self.pontos - 1)
        if largura > 1e-6 * max(self.inicio_hz, 1.0):
            raise ErroVNA("amostras_cw exige frequência fixa: a grade atual "
                          "cobre %.3f Hz. Chame define_cw antes." % largura)
        _, s11, s21 = self._com_retentativa(lambda: self._le_n(int(n)))
        return s11, s21


# ------------------------------------------------------------------------ CLI
def _fmt_db(z):
    m = np.abs(z)
    with np.errstate(divide="ignore", invalid="ignore"):
        return 20 * np.log10(m)


def _cli(argv):
    import argparse

    ap = argparse.ArgumentParser(
        prog="mwflow.vna.litevna",
        description="Driver de linha de comando do LiteVNA64.")
    ap.add_argument("--porta", default=None, help="porta serial")
    ap.add_argument("--info", action="store_true", help="identificação")
    ap.add_argument("--varre", nargs=3, metavar=("INICIO", "FIM", "PONTOS"),
                    help="uma varredura, em Hz")
    ap.add_argument("--cw", metavar="FREQ", type=float,
                    help="modo de frequência fixa, em Hz")
    ap.add_argument("--n", type=int, default=500,
                    help="amostras no modo CW (padrão 500)")
    ap.add_argument("--media", type=int, default=1,
                    help="varreduras a promediar (padrão 1)")
    a = ap.parse_args(argv)

    vna = LiteVNA(a.porta)
    try:
        if a.info or not (a.varre or a.cw):
            i = vna.info()
            print("porta            :", i["porta"])
            print("deviceVariant    : 0x%02X %s"
                  % (i["variante"],
                     "(NanoVNA-V2 / LiteVNA)" if i["variante"] == p.VARIANTE_V2
                     else "(variante desconhecida)"))
            print("protocolVersion  :", i["protocolo"])
            print("hardwareRevision :", i["hardware"])
            print("firmware         :", i["firmware"])

        if a.varre:
            ini, fim, pts = float(a.varre[0]), float(a.varre[1]), int(a.varre[2])
            vna.define_varredura(ini, fim, pts)
            t0 = time.time()
            f, s11, s21, dp = vna.varre_media(a.media)
            dt = time.time() - t0
            faltam = int(np.isnan(s11).sum())
            print()
            print("varredura: %.3f a %.3f MHz, %d pontos, %d média(s), %.2f s "
                  "(%.0f pontos/s)"
                  % (ini / 1e6, fim / 1e6, pts, a.media, dt,
                     pts * a.media / dt))
            print("pontos sem dado: %d" % faltam)
            print()
            print("  freq (MHz)   S11 (dB)   S21 (dB)")
            passo = max(1, pts // 6)
            for k in range(0, pts, passo):
                print("  %10.3f   %8.2f   %8.2f"
                      % (f[k] / 1e6, _fmt_db(s11[k]), _fmt_db(s21[k])))

        if a.cw:
            vna.define_cw(a.cw)
            t0 = time.time()
            s11, s21 = vna.amostras_cw(a.n)
            dt = time.time() - t0
            d11, d21 = _fmt_db(s11), _fmt_db(s21)
            print()
            print("CW em %.6f MHz: %d amostras em %.2f s (%.0f amostras/s)"
                  % (a.cw / 1e6, a.n, dt, a.n / dt))
            print("S11: %.3f dB  desvio %.4f dB" % (np.nanmean(d11),
                                                    np.nanstd(d11)))
            print("S21: %.3f dB  desvio %.4f dB" % (np.nanmean(d21),
                                                    np.nanstd(d21)))
    except ErroVNA as e:
        print("ERRO:", e, file=sys.stderr)
        return 1
    finally:
        vna.fecha()
    return 0


if __name__ == "__main__":
    sys.exit(_cli(sys.argv[1:]))
